File: lib/activity-log.ts
import { prisma } from './prisma'

export interface LogActivityParams {
  action: string
  description: string
  userId?: string
  userName?: string
  userRole?: string
  targetId?: string
  targetType?: 'submission' | 'user' | 'review'
  metadata?: Record<string, any>
}

export async function logActivity(params: LogActivityParams) {
  try {
    await prisma.activityLog.create({
      data: {
        action: params.action,
        description: params.description,
        userId: params.userId,
        userName: params.userName,
        userRole: params.userRole,
        targetId: params.targetId,
        targetType: params.targetType,
        metadata: params.metadata ? JSON.stringify(params.metadata) : null,
      },
    })
  } catch (error) {
    console.error('Failed to log activity:', error)
    // Don't throw - logging failures shouldn't break the app
  }
}

// Helper to get recent logs for admin dashboard
export async function getRecentLogs(limit: number = 50) {
  return prisma.activityLog.findMany({
    take: limit,
    orderBy: { createdAt: 'desc' },
  })
}

// Helper to get logs for a specific user
export async function getUserLogs(userId: string, limit: number = 50) {
  return prisma.activityLog.findMany({
    where: { userId },
    take: limit,
    orderBy: { createdAt: 'desc' },
  })
}

// Helper to get logs for a specific submission
export async function getSubmissionLogs(submissionId: string) {
  return prisma.activityLog.findMany({
    where: {
      targetId: submissionId,
      targetType: 'submission',
    },
    orderBy: { createdAt: 'desc' },
  })
}

File: backend/internal/services/activity_log.go
package services

import (
	"encoding/json"

	"github.com/adzzatxperts/backend/internal/database"
	"github.com/adzzatxperts/backend/internal/models"
	"github.com/google/uuid"
)

type LogActivityParams struct {
	Action      string
	Description string
	UserID      *uuid.UUID
	UserName    *string
	UserRole    *string
	TargetID    *uuid.UUID
	TargetType  *string
	Metadata    map[string]interface{}
}

func LogActivity(params LogActivityParams) error {
	var metadataJSON *string
	if params.Metadata != nil {
		bytes, err := json.Marshal(params.Metadata)
		if err == nil {
			str := string(bytes)
			metadataJSON = &str
		}
	}

	log := models.ActivityLog{
		Action:      params.Action,
		Description: params.Description,
		UserID:      params.UserID,
		UserName:    params.UserName,
		UserRole:    params.UserRole,
		TargetID:    params.TargetID,
		TargetType:  params.TargetType,
		Metadata:    metadataJSON,
	}

	return database.DB.Create(&log).Error
}

func GetRecentLogs(limit int) ([]models.ActivityLog, error) {
	var logs []models.ActivityLog
	err := database.DB.
		Order("created_at DESC").
		Limit(limit).
		Find(&logs).Error
	return logs, err
}
